src/audio/codec_adapter/codec_adapter.c

```diff
@@ -406,6 +406,13 @@ static int codec_adapter_copy(struct comp_dev *dev)
 			break;
 		}
 
+		/* TODO: remove magic number and use possible maximum output per
+		 * decoding algorithm type
+		 * 9216 is maximum frame length bytes for mp3 at 192Khz, 24bits,
+		 */
+		if (local_buff->stream.free < 9217)
```

Comment:
@thanks Iulia for the comment. What I actually plan to do is to create a generic function that given a codec and its parameters it will return the maximum amount of the output data.

Something like `code_adapter_get_max_bytes`? I just couldn't find a good name. This work is still RFC. We won't submit the code like that.

Reply:
A generic function sounds good. 
Regarding the naming, I think it's ok, but I'm not good with names either :)

Reply:
Fwiw the 9212 should also be a topology 2.0 attribute (as this can also be enforced in any MP3 topology)

Reply:
@lgirdwood this could be an option. It looks like at least for mp3 the number of samples resulted after decoding one frame is always the same 1152 (9217 = 1152 samples * 2 channels * 4 bytes / sample + 1). If this holds true for all the decoding algorithms then getting this via topology could be an option.

Reply:
it's not true for the lower frequencies, the result is 576 samples.

More generically, there's actually a need to decouple input frames and output 'granules'. Sometimes frames are consumed without generating any data (e.g. to check for sync words or to build a history), and conversely on stream close there may still be data in a history buffer that needs to be flushed even with no incoming data. In general though the output granularity is constant and defined by the filter banks.

Reply:
@plbossart 

![image](https://user-images.githubusercontent.com/62316/112663922-0cf11f00-8e62-11eb-99a4-69a4d959ff7b.png)

Reply:
I think it would be a grave mistake to use a single period. 

if the scheduling is constrained by the output size, then it may happen than you have to decode multiple frames in the allocated time. In our past integration, the decoder was part of a different pipeline and scheduled with a lower priority, so that there were no peaks in the MCPS usage. I really think you have to decouple the codec_adapter from the dai scheduling....

Reply:
@plbossart Definitely! That was my next step. I try to do it gradually and also have something that's working for our releases. Customers want to look at the DSP even if it isn't the final code.
I will try to see how we satisfy both requirements. In time! :)

- the only thing that's *definitely* out of place right now is the m4 topology.

For the rest of the code:
- existing DAP support (based on PCM frames, should still work)
- no other functionality is impacted.

Next steps:
(1) decouple codec adapter from DAI scheduling. In fact here there is no clear line where should I do the decoupling.

Because codec_adapter has an output buffer that holds the decoded samples and DMA needs to copy based on DAI pace
each period "period_bytes" of data.
So, each DAI period I need to call codec_adapter_copy() to push data from codec_adapter local buffer to DMA buffer. 

Also, the Host need to push data to codec_adapter based on its own pace. So, Host should copy one "compress period" of data each time the DAI has rendered one mp3 frame.

So, in simple terms: 
- Host should copy one mp3 compressed frame each 0.026 sec (compress period will be 26ms).
- DAI should copy SAMPLE_RATE/1000 sampes each 1ms (compress period will be 1ms).

Also, we noticed that we could process around 3 MP3 frames with a single pipline, so there is enough room for our simple sceario to work.

Reply:
I think you should really have 3 pipelines
a) dai with mixer
b) PCM for notifications/beeps that may occur at any time
c) compressed. That pipe may run at a lower priority or larger granularity than the DAI period. The last thing you want it to try and complete one decode at a 26x speed, you have to spread the work across multiple dai periods otherwise you *will*  underrun when your system is loaded.

IIRC in all implementations, the DAI typically works with O(ms) buffers, and the compressed pipeline can work with much larger frames. In fact if you look at newer codecs they will generate 4k or 8k samples in one frame. So you need some sort of ring buffer to adjust between codec and dai.

Reply:
And I forget that you need an SRC on your compressed pipeline if you want to support all frequencies.

Reply:
Nice suggestion.